webui/react/src/hooks/useStreamingProvider.tsx

```diff
@@ -0,0 +1,62 @@
+import { forEach } from 'lodash';
+import React, { createContext, useCallback, useEffect, useMemo } from 'react';
+
+import { serverAddress } from 'routes/utils';
+import { Streamable, StreamContent, StreamEntityMap } from 'services/stream';
+import { Stream } from 'services/stream/stream';
+import projectStore, { mapStreamProject } from 'stores/projects';
+
+type UserStreamingContext = {
+  stream?: Stream;
+};
+
+export const Streaming = createContext<UserStreamingContext>({
+  stream: undefined,
+});
+
+export const StreamingProvider: React.FC<React.PropsWithChildren> = ({ children }) => {
```

Comment:
i'm not a fan of this setup. having the provider update the stores will mean that the logic here will inflate as more stores get streaming updates and more types of events are supported. furthermore, operation of the stores is currently separate from the react lifecycle -- passing the current stream around by context breaks that expectation.

consider instead having this module construct and dispatch events that stores can listen to to operate on themselves. maintain the stream by only having it open when there are active subscribers.

Reply:
Hi @ashtonG , thanks for your advice, I've replaced the `StreamingProvider` with the pub/sub pattern as you suggested. 
Also I'm not sure about defining socket url as `${serverAddress().replace('http', 'ws')}/stream`, I wonder what would be the best practice to access websocket url?

Reply:
that should be fine -- using a string as the pattern to replace in a string will cause the first occurrence to be replaced, so places where the server address contains multiple instances of `http` like our netlify preview should be unaffected. we should make sure that we support `wss` (websockets secure) protocol, but I think if we didn't that'd trigger some sort of mixed content warning.

i guess ideally you'd parse the url and set the protocol specifically?
```tsx
const socketUrl = new URL(serverAddress())
socketUrl.path = 'stream'
socketUrl.protocol = socketUrl.protocol.replace('http', 'ws');
```